osh/cmd_parse.py

```diff
@@ -2080,11 +2092,11 @@ def ParsePipeline(self):
     negated = False
 
     # For blaming failures
-    pipeline_spid = runtime.NO_SPID
+    pipeline_op = None  # type: Token
```

Comment:
Hmm I agree that `!` should be the `Token left_tok` if present

But I'm not sure about `|` otherwise  -- I think it should be the left-most location of the command that the pipeline starts with